### PR TITLE
cgroupstats: further unify cgroupstats getters.

### PR DESCRIPTION
**Pull request by klihub**
This patchset unifies parsing in the complex cgroupstats getters. It
  - adds a configurable parser for parsing multiple single-line entries from a file
  - reworks GetBlkioThrottleBytes and GetNumaStats using the above function
  - adds support for delayed evaluation of (expensive) arguments for log.Debug()

### CODE REVIEW

File: pkg/sysfs/parsers.go [hunk 1]
```diff
@@ -115,6 +133,65 @@ func parseNumeric(path, value string, ptr interface{}) error {
...
+		vmap.(map[string]uint)[key] = uint(unum * uint64(unit))
+	case map[string]uint8:
+		unum, err = strconv.ParseUint(numstr, 0, 8)
+		vmap.(map[string]uint8)[key] = uint8(unum * uint64(unit))
+	case map[string]uint16:
+		unum, err = strconv.ParseUint(numstr, 0, 16)
+		vmap.(map[string]uint16)[key] = uint16(unum * uint64(unit))
+	case map[string]uint32:
+		unum, err = strconv.ParseUint(numstr, 0, 32)
+		vmap.(map[string]uint32)[key] = uint32(unum * uint64(unit))
```
Comment (ipuustin) [inline pkg/sysfs/parsers.go]: Why would we want anything other than `{int64,uint64,float64}` as internal result values? When I see code such as `uint8(unum * uint64(unit))` I start thinking that this is going overflow.

File: pkg/cgroups/cgroupstats.go [hunk 2]
```diff
@@ -368,72 +379,73 @@ func GetNumaStats(cgroupPath string) (NumaStat, error) {
...
-		for _, nodeEntry := range split[1:] {
-			nodeamount := strings.Split(nodeEntry, "=")
-			if len(nodeamount) != 2 {
-				return NumaStat{}, fmt.Errorf("error parsing file %s", entry)
+	path := filepath.Join(cgroupPath, cgroupEntry)
+	err := sysfs.ParseFileByLines(path,
+		// split line to fields by spaces
+		func(line string) ([]*sysfs.Field, error) {
+			split := strings.Split(line, " ")
+			fields := make([]*sysfs.Field, 0, len(split))
```
[the remark this answers is not in the document]
Comment (klihub) [reply pkg/cgroups/cgroupstats.go]: It's an idiomatic pattern I picked up for slice-construction in a loop: voluntarily pay the price of a single call per iteration to the fast path of the builtin append instead of ending up hard-to find bugs with trailing zero-value entries in case we ever break out early form the loop and (try to) return the items collected so far.

But you're right, with this simple loop it cannot happen. Updated code accordingly.

File: pkg/sysfs/parsers.go [hunk 3]
```diff
@@ -166,3 +246,74 @@ func ParseFileEntries(path string, values map[string]interface{}, pickFn PickEnt
...
+
+	for _, line := range strings.Split(string(data), "\n") {
+		fields, err := splitFn(line)
+		switch {
+		case err == ErrSkip:
+			continue
+		case err != nil:
+			return nil
+		}
+		sort.Slice(fields, func(i, j int) bool { return fields[i].Index <= fields[j].Index })
```
[the remark this answers is not in the document]
Comment (klihub) [reply pkg/sysfs/parsers.go]: Added a comment explaining why what the caller-provided Index indicates and why we sort.

File: pkg/sysfs/parsers.go (hunk 1 above)
Comment (klihub) [reply pkg/sysfs/parsers.go]: Valid point. Updated to use highest-precision internal representations and added checks for trivial overflow/underflow of the numeric part without the unit.

**Comment by klihub** [comment]
> I'm bit on the fence with the generalization of the parsing logic... on the other hand it's good to have reusable code and there may well be more files which we are parsing in the future, but on the other hand the parsing logic starts to be pretty abstract and tricky to follow.

I'm pretty sure that with a bit of head scratching it's possible to come up with a simpler yet equally generic and customizable parser for line-oriented files. It's just not something I'd personally want to spend too much time with right now. If someone has a much better alternative, I'm glad to take it though.

File: pkg/log/log.go [hunk 4]
```diff
@@ -55,6 +56,62 @@ type Logger interface {
...
 
+// Delayed implement delayed evaluation, useful for lowering the overhead suppressed log.Debug.
+type Delayed interface {
+	String() string
+}
+
+// delay implements Delayed.
+type delay struct {
+	o interface{}
+	Delayed
```
Comment (marquiz) [inline pkg/log/log.go]: You could drop this `Delayed`

File: pkg/sysfs/parsers.go [hunk 5]
```diff
@@ -115,6 +134,145 @@ func parseNumeric(path, value string, ptr interface{}) error {
...
+		vm[key] = num
+	case map[string]uint64:
+		typeName, max := "uint64", uint64(math.MaxUint64)
+		if unum, err = parseUintWithUnit(path, numstr, unit); err != nil {
+			return err
+		}
+		if unum > max {
+			return sysfsError(path, "numeric value %s overflows %s", value, typeName)
+		}
+		vm[key] = unum
```
Comment (marquiz) [inline pkg/sysfs/parsers.go]: There's a lot of duplicate code. Just a thought: could this be simplified by doing the limits checking in `parseUintWithUnit()` / `parseIntWithUnit()`? Either by giving the number of bits or min/max limits as arguments. WDYT?

Comment (klihub) [reply pkg/sysfs/parsers.go]: Yes, that's probably a good idea. I'll change it accordingly.

File: pkg/sysfs/parsers.go [hunk 6]
```diff
@@ -166,3 +327,76 @@ func ParseFileEntries(path string, values map[string]interface{}, pickFn PickEnt
...
+		for idx, f := range fields {
+			ptr, err := pickFn(idx, f)
+			switch {
+			case err == ErrSkip:
+				continue
+			case err != nil:
+				return err
+			}
+
+			switch ptr.(type) {
```
Comment (marquiz) [inline pkg/sysfs/parsers.go]: Would be possible to use the `p := ptr.(type)` construct here, agani?

**Comment by marquiz** [comment]
> I'm bit on the fence with the generalization of the parsing logic... on the other hand it's good to have reusable code and there may well be more files which we are parsing in the future, but on the other hand the parsing logic starts to be pretty abstract and tricky to follow.

After finally having time to read through the last patch, I'm inclined to second @ipuustin here. The code starts to be hard to follow, especially at this time on Friday evening 😄

File: pkg/sysfs/parsers.go (hunk 6 above)
Comment (klihub) [reply pkg/sysfs/parsers.go]: I wonder how long it will take to get this into my thick skull... if ever. ;-)
I'll update the code.

**Comment by klihub** [comment]
Closing this for now...
